### PR TITLE
[AArch64] Define high bits of FPR and GPR registers (take 2)

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
This is a step towards enabling subreg liveness tracking for AArch64, which requires that registers are fully covered by their subregisters, as covered here #109797.

There are several changes in this patch:

* AArch64RegisterInfo.td and tests: Define the high bits like B0_HI, H0_HI, S0_HI, D0_HI, Q0_HI. Because the bits must be defined by some register class, this added a register class which meant that we had to update 'magic numbers' in several tests.

  The use of ComposedSubRegIndex helped 'compress' the number of bits required for the lanemask. The correctness of the masks is tested by an explicit unit tests.

* LoadStoreOptimizer: previously 'HasDisjunctSubRegs' was only true for register tuples, but with this change to describe the high bits, a register like 'D0' will also have 'HasDisjunctSubRegs' set to true (because it's fullly covered by S0 and S0_HI). The fix here is to explicitly test if the register class is one of the known D/Q/Z tuples.

* TableGen: The handling of the isArtificial flag was entirely broken. Skipping out too early from some of the loops led to incorrect internal representation of the (sub)register(index) hierarchy, and thus resulted in incorrect TableGen info.

(This follows on from #114263 which was mysteriously closed.)

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64RegisterInfo.td [hunk 1]
```diff
@@ -20,33 +20,39 @@ class AArch64Reg<bits<16> enc, string n, list<Register> subregs = [],
 
 let Namespace = "AArch64" in {
   // SubRegIndexes for GPR registers
-  def sub_32 : SubRegIndex<32>;
-  def sube64 : SubRegIndex<64>;
-  def subo64 : SubRegIndex<64>;
-  def sube32 : SubRegIndex<32>;
-  def subo32 : SubRegIndex<32>;
+  def sub_32   : SubRegIndex<32,  0>;
+  def sub_32_hi: SubRegIndex<32, 32>;
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/AArch64RegisterInfo.td:24]: Comment the artificial cases? Could we use an explicit marker for these?

Comment (sdesmalen-arm) [reply llvm/lib/Target/AArch64/AArch64RegisterInfo.td]: I've added a comment to describe the artificial cases. I'm not sure what you mean with an explicit marker though?

**Comment by sdesmalen-arm** [comment]
@Rin18 would you be happy to have a look at the `V1-clear-upper-regs.s` test to see if there's any issues with the generated output? I'm not sure if there is a hidden bug in there (it seems to iterate through all subregs, including artificial ones, which may affect the output)

**Comment by Rin18** [comment]
> @Rin18 would you be happy to have a look at the `V1-clear-upper-regs.s` test to see if there's any issues with the generated output? I'm not sure if there is a hidden bug in there (it seems to iterate through all subregs, including artificial ones, which may affect the output)

I took a look and compared the MCA output with and without your patch. The first thing that jumped to my attention is that your patch adds dependencies on super registers. If you run the V1-clear-upper-regs.s test with `-debug` and look at the output for the GPR32-bit test, you'll see a lot of these lines:
```
[PRF] Found a dependent use of Register X0 (defined by instruction #
```
Similarly for the other super registers in the remaining tests in that file. 

I've not been able to find the exact cause. I think this has something to do with the addition of `*_HI` registers and the fact that super registers are now described as `*sub + *sub_hi`.

**Comment by sdesmalen-arm** [comment]
I'm not entirely sure how llvm-mca works, but I think the issue is that the code adds dependences for all sub-registers of a register, but when `clearsSuperRegisters` evalutes to true it clears dependences for super-registers without considering any of the sibling `_hi` registers. For example; a write to `w0` also clears a dependence on `x0`, but the dependence on artificial register `w0_hi` should also be cleared. It seems easiest not to consider artificial registers in the computation to start with.

**Comment by Rin18** [comment]
>  but I think the issue is that the code adds dependences for all sub-registers of a register, but when clearsSuperRegisters evalutes to true it clears dependences for super-registers without considering any of the sibling _hi registers

That sounds right. Before your patch, there were no `_hi` registers to consider when updating dependencies. Now that artificial registers are added it makes sense to skip them. Looks like that fixed the dependencies created in the MCA `*clear-upper-regs.s` tests, thanks!

**Review by davemgreen — COMMENTED** [review]
The general register changes sound to me.

**Review by davemgreen — APPROVED** [review]
Thanks. If there are no objections to the name "Artificial", then LGTM.

**Comment by sdesmalen-arm** [comment]
> Thanks. If there are no objections to the name "Artificial", then LGTM.

Thanks @davemgreen! The name Artificial is already established in `TableGen` and `include/llvm/Target/Target.td`, I've just reused the name here.

File: llvm/lib/Target/AArch64/AArch64RegisterInfo.td (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AArch64/AArch64RegisterInfo.td]: I mean an isArtificial bit on SubRegIndex

Comment (sdesmalen-arm) [reply llvm/lib/Target/AArch64/AArch64RegisterInfo.td]: `isArtificial` is a property of a Register or a RegisterClass, but not a SubRegIndex.

**Comment by zeroomega** [comment]
Hi, 

We noticed a large clang/llvm test time regression on our clang linux-arm64 and mac-arm64 builders after this patch is landed.
The test step took around 18mins before this patch was landed see: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-arm64/b8731309118845812913/overview (please expand the 'clang' step).

The test step took around 41mins after this change was landed. see: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-arm64/b8731304524561843377/overview

Could you take a look?
If it takes some time to fix, could you revert your change first?

**Comment by PiJoules** [comment]
Also this really affects our mac arm64 builders from 2.8hrs to do a full build and run tests (https://ci.chromium.org/ui/p/fuchsia/builders/prod/clang-mac-arm64/b8731291539321484417/overview) to >5hrs for a full build and run tests (https://ci.chromium.org/ui/p/fuchsia/builders/prod/clang-mac-arm64/b8731291539321484417/overview).

**Comment by sdesmalen-arm** [comment]
Thanks for letting me know. From what I can quickly see there is only a time difference for the `runtime` tests. The `clang`, `polly`, `llvm` and `lld` tests run in about the same time. @PiJoules the build you linked to was interrupted while building the `check-clang` target, which for the other builds pointed to by @zeroomega wasn't any slower. Perhaps I'm missing something here though.
I do see various other patches in the blamelist, has it been confirmed that this patch is the cause of the increased testing time?

**Comment by zeroomega** [comment]
> Thanks for letting me know. From what I can quickly see there is only a time difference for the `runtime` tests. The `clang`, `polly`, `llvm` and `lld` tests run in about the same time. @PiJoules the build you linked to was interrupted while building the `check-clang` target, which for the other builds pointed to by @zeroomega wasn't any slower. Perhaps I'm missing something here though. I do see various other patches in the blamelist, has it been confirmed that this patch is the cause of the increased testing time?

The patch is confirmed through bisecting. 

https://ci.chromium.org/b/8730624321112150225 builds e52238b59f250aef5dc0925866d0308305a19dbf and the test time looks normal.

https://ci.chromium.org/b/8730630997932921601 builds c1c68baf7e0fcaef1f4ee86b527210f1391b55f6 and the test time regressed. c1c68baf7e0fcaef1f4ee86b527210f1391b55f6 (https://github.com/llvm/llvm-project/pull/114827) is 1 commit after e52238b59f250aef5dc0925866d0308305a19dbf


For stage2 builders, I didn't do a bisecting since they uses LTO and will take too much time. But
https://ci.chromium.org/ui/p/fuchsia/builders/prod/clang-linux-arm64/b8730794949594129265/overview is from 2 days ago (of course after this patch was landed). It took 2.5hrs for building clang and runtimes. and 2.1hrs for running all the tests. 
https://ci.chromium.org/ui/p/fuchsia/builders/prod/clang-linux-arm64/b8731348039888781153/overview is from 11/13 (before this patch was landed). It took 43mins to build clang and runtimes and 18 minutes to run the tests. If you feel it is necessary to perform a before/after stage2 build test on build time, though we feel it is unnecessary since the bisecting on the stage1 build already shows it clearly,, we can do that, it will take about 6 hours. 

I suspect the patch causes a general performance regression of binaries built by clang. Because the stage1 clang build is using the host toolchian, so the build time wasn't affected. But clang, lld and runtimes tests are using the freshly built clang, they are affected. The stage2 builder is built using the stage1 toolchain (which has the performance bug), so the build time also regressed. 

Let us know if you need additional information.

**Comment by vitalybuka** [comment]
This patch makes libcxx tests to timeouts here
https://lab.llvm.org/buildbot/#/builders/85

Many tests slowed down 10x or more.

**Pull request merged**